src/server/hotkey_collector.cpp

```diff
@@ -0,0 +1,265 @@
+// Copyright (c) 2017, Xiaomi, Inc.  All rights reserved.
+// This source code is licensed under the Apache License Version 2.0, which
+// can be found in the LICENSE file in the root directory of this source tree.
+
+#include "hotkey_collector.h"
+
+#include "base/pegasus_key_schema.h"
+#include "base/pegasus_rpc_types.h"
+#include <math.h>
+
+namespace pegasus {
+namespace server {
+
+DSN_DEFINE_int32("pegasus.server",
+                 hotkey_collector_max_work_time,
+                 150,
+                 "after hotkey_collector_max_work_time the collection will stop automatically");
+
+DSN_DEFINE_int32("pegasus.server",
+                 data_capture_hash_bucket_num,
+                 37,
+                 "the number of data capture hash buckets");
+
+inline bool hotkey_collector::_is_ready_to_detect()
+{
+    return (_collector_state.load() == collector_state::STOP ||
+            _collector_state.load() == collector_state::FINISH);
+}
+
+/*static*/ int hotkey_collector::data_hash_method(const std::string &data)
```

Comment:
```suggestion
/*static*/ int hotkey_collector::get_bucket_id(const std::string &data)
```

"_method" is meaningless. I would not see what it supposes to mean if I didn't look at the code.